File: web/src/services/ipam.ts
import request from '../utils/request'
import type { PageResult } from './audit'

export interface SubnetItem {
  id: number
  name: string
  cidr: string
  gateway: string
  vlan: number
  total_ips: number
  used_ips: number
  usage_percent: number
  comment: string
}

export interface IPAddressItem {
  id: number
  subnet_id: number
  address: string
  status: string
  asset_id: number | null
  hostname: string
  comment: string
}

export interface CreateSubnetParams {
  name: string
  cidr: string
  gateway?: string
  vlan?: number
  comment?: string
}

export function getSubnets(params?: { page?: number; page_size?: number }): Promise<PageResult<SubnetItem>> {
  return request.get('/ipam/subnets', { params })
}

export function createSubnet(data: CreateSubnetParams): Promise<SubnetItem> {
  return request.post('/ipam/subnets', data)
}

export function updateSubnet(id: number, data: Partial<CreateSubnetParams>): Promise<void> {
  return request.put(`/ipam/subnets/${id}`, data)
}

export function deleteSubnet(id: number): Promise<void> {
  return request.delete(`/ipam/subnets/${id}`)
}

export function getSubnetIPs(id: number, params?: { status?: string; page?: number; page_size?: number }): Promise<PageResult<IPAddressItem>> {
  return request.get(`/ipam/subnets/${id}/ips`, { params })
}

export function allocateIP(ipId: number, data: { asset_id?: number; hostname?: string; comment?: string }): Promise<void> {
  return request.post(`/ipam/ips/${ipId}/allocate`, data)
}

export function releaseIP(ipId: number): Promise<void> {
  return request.post(`/ipam/ips/${ipId}/release`)
}

File: internal/dto/ipam.go
package dto

// CreateSubnetRequest 创建网段请求
type CreateSubnetRequest struct {
	Name    string `json:"name" binding:"required,max=128"`
	CIDR    string `json:"cidr" binding:"required"`
	Gateway string `json:"gateway"`
	VLAN    int    `json:"vlan"`
	Comment string `json:"comment"`
}

// UpdateSubnetRequest 更新网段请求
type UpdateSubnetRequest struct {
	Name    string `json:"name"`
	Gateway string `json:"gateway"`
	VLAN    int    `json:"vlan"`
	Comment string `json:"comment"`
}

// SubnetResponse 网段响应
type SubnetResponse struct {
	ID       uint    `json:"id"`
	Name     string  `json:"name"`
	CIDR     string  `json:"cidr"`
	Gateway  string  `json:"gateway"`
	VLAN     int     `json:"vlan"`
	TotalIPs int     `json:"total_ips"`
	UsedIPs  int     `json:"used_ips"`
	UsagePercent float64 `json:"usage_percent"`
	Comment  string  `json:"comment"`
}

// AllocateIPRequest 分配 IP 请求
type AllocateIPRequest struct {
	AssetID  *uint  `json:"asset_id"`
	Hostname string `json:"hostname"`
	Comment  string `json:"comment"`
}

// IPAddressResponse IP 地址响应
type IPAddressResponse struct {
	ID       uint    `json:"id"`
	SubnetID uint    `json:"subnet_id"`
	Address  string  `json:"address"`
	Status   string  `json:"status"`
	AssetID  *uint   `json:"asset_id"`
	Hostname string  `json:"hostname"`
	Comment  string  `json:"comment"`
}
